### PR TITLE
Graceful shutdown refactor

### PR DESCRIPTION
**Pull request by biandratti**
A new graceful shutdown mechanism was implemented that waits for existing connections to drain while continuing to accept new connections for a specified period, thereby allowing the load balancer to detect when the proxy has initiated the shutdown process.

## Shutdown sequence

SIGTERM / SIGINT
│
├─▶ phase = Draining                (server.rs)  /ready = 503
│     accept loops keep running
│     sleep drain_delay_secs (second signal skips)
│
├─▶ phase = Stopping
│     │
│     ├─▶ accept loops              wait_for(Stopping) → break
│     ├─▶ config-watcher            wait_for(Stopping) → break
│     ├─▶ ebpf-reconnect            wait_for(Stopping) → break
│     └─▶ connections               graceful_shutdown() then wait_for_drain
│
└─▶ observability server            stopped by main.rs after run() returns

### CODE REVIEW

File: huginn-proxy-lib/src/telemetry/readiness.rs [hunk 1]
```diff
@@ -3,27 +3,77 @@
...
 
     /// Mark the proxy as ready to accept traffic (`/ready` -> 200).
     pub fn mark_ready(&self) {
-        self.0.store(true, Ordering::Release);
+        self.0.draining.store(false, Ordering::Release);
+        self.0.ready.store(true, Ordering::Release);
     }
 
     /// Mark the proxy as not ready (`/ready` -> 503), e.g. during graceful shutdown.
     pub fn mark_not_ready(&self) {
```
Comment (ganna-loyik) [inline huginn-proxy-lib/src/telemetry/readiness.rs]: it's not used

File: huginn-proxy-lib/src/telemetry/readiness.rs [hunk 2]
```diff
@@ -3,27 +3,77 @@
...
-    /// Create a new handle in the not-ready state.
+    /// Create a new handle in the not-ready state (`proxy_starting`).
     pub fn new() -> Self {
         Self::default()
     }
 
     /// Mark the proxy as ready to accept traffic (`/ready` -> 200).
     pub fn mark_ready(&self) {
-        self.0.store(true, Ordering::Release);
+        self.0.draining.store(false, Ordering::Release);
```
Comment (ganna-loyik) [inline huginn-proxy-lib/src/telemetry/readiness.rs]: maybe add `if draining { return; }`?

File: huginn-proxy-lib/src/proxy/transport/timeout_helper.rs [hunk 3]
```diff
@@ -1,25 +1,70 @@
...
+                        debug!(?peer, reason = %e, "connection ended");
+                    }
+                }
+                return;
+            }
+            _ = shutdown_rx.wait_for(|phase| phase.is_stopping()), if !goaway_sent => {
+                serve_fut.as_mut().graceful_shutdown();
+                goaway_sent = true;
+            }
+            _ = tokio::time::sleep_until(deadline) => {
```
Comment (ganna-loyik) [inline huginn-proxy-lib/src/proxy/transport/timeout_helper.rs]: can you use `pingora_timeout`? it is optimised better when there are lots of connections

File: huginn-proxy-lib/tests/proxy/shutdown.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    let _ = tx.send(());
+    assert!(waiter.await.is_ok());
+}
+
+#[tokio::test]
+async fn wait_for_drain_times_out_with_connections_still_active() {
+    let (_tx, rx) = watch::channel(());
+    let active = Arc::new(AtomicUsize::new(3));
+    wait_for_drain(rx, active, 0).await;
+}
```
Comment (ganna-loyik) [inline huginn-proxy-lib/tests/proxy/shutdown.rs:48]: can you add a test to check that phase 1 accepts while `/ready` is 503?

File: huginn-proxy-lib/src/telemetry/readiness.rs (hunk 2 above)
Comment (biandratti) [reply huginn-proxy-lib/src/telemetry/readiness.rs]: I have change and simplify the readiness process again

File: huginn-proxy-lib/src/telemetry/readiness.rs (hunk 1 above)
Comment (biandratti) [reply huginn-proxy-lib/src/telemetry/readiness.rs]: right, done

File: huginn-proxy-lib/src/proxy/transport/timeout_helper.rs (hunk 3 above)
Comment (biandratti) [reply huginn-proxy-lib/src/proxy/transport/timeout_helper.rs]: good catch, yes, the original idea of using pingora timeout was to be prepare to split read and write timeout in the forwarding. Thanks for this review.

File: huginn-proxy-lib/tests/proxy/shutdown.rs (hunk 4 above)
Comment (biandratti) [reply huginn-proxy-lib/tests/proxy/shutdown.rs:48]: good point, I have added corner cases for different phases. Thanks!

**Review by ganna-loyik — APPROVED** [review]
